File: mac/Sources/ZislaCore/AIModels.swift
import Foundation

/// Supported AI providers. `rawValue` is used as a stable identifier in CLI output and persistence.
public enum AIProvider: String, Codable, CaseIterable, Sendable {
    case claude
    case codex
    case gemini
    case grok
    case gpt
    case copilot
    case kimi
    case qwen
    case coder
    case trae
    case opencode
    case harness
    case doubao

    /// Normalises case and common aliases; returns nil for unknown tokens, letting the caller report the error.
    public init?(token: String) {
        switch token.lowercased() {
        case "claude", "claude-code", "claude-cli", "claude-desktop", "anthropic-claude":
            self = .claude
        case "codex", "openai-codex", "codex-cli", "codex-desktop", "codex-app":
            self = .codex
        case "gemini", "google-gemini", "gemini-cli", "gemini-code-assist":
            self = .gemini
        case "grok", "grok-cli", "xai", "x-ai":
            self = .grok
        case "gpt", "openai", "chatgpt", "chat-gpt", "openai-gpt":
            self = .gpt
        case "copilot", "github-copilot", "github copilot", "copilot-cli", "copilot-chat", "github.copilot-chat":
            self = .copilot
        case "kimi", "kimi-code", "kimi-code-cli", "kimi-vscode", "moonshot-kimi", "moonshot-ai.kimi-code":
            self = .kimi
        case "qwen", "tongyi", "qwen-code", "qwen-code-cli", "qwen-vscode":
            self = .qwen
        case "coder", "qwen-coder", "qoder", "qoder-cli", "qoderwork", "qoder-work",
             "qoderworkcn", "qoderwork-cn", "qoderwork cn", "qoderwake", "qoder-wake":
            self = .coder
        case "trae", "trae-work", "traework", "trae-work-cn", "trae-solo", "trae-solo-cn", "trae-cn":
            self = .trae
        case "opencode", "open-code", "open_code":
            self = .opencode
        case "harness", "harnext", "harnext-cli", "harness-cli":
            self = .harness
        case "doubao", "豆包":
            self = .doubao
        default: return nil
        }
    }
}

public enum AIProgressStatus: String, Codable, Sendable {
    case queued
    case running
    case blocked
    case error
    case succeeded
    case failed

    public var isActive: Bool {
        switch self {
        case .queued, .running, .blocked, .error:
            true
        case .succeeded, .failed:
            false
        }
    }

    public var noticeKind: NoticeKind {
        switch self {
        case .queued, .running: .info
        case .blocked: .warning
        case .error, .failed: .error
        case .succeeded: .success
        }
    }
}

/// Single AI task progress record. `progress == nil` indicates indeterminate progress (spinner).
public struct AIProgressTask: Codable, Equatable, Sendable, Identifiable {
    public var id: String
    public var provider: AIProvider
    public var title: String
    public var detail: String?
    public var progress: Double?
    public var status: AIProgressStatus
    public var updatedAt: Date
    public var sessionURL: URL?
    public var effort: String?
    public var startedAt: Date?

    public init(
        id: String,
        provider: AIProvider,
        title: String,
        detail: String? = nil,
        progress: Double?,
        status: AIProgressStatus = .running,
        updatedAt: Date,
        sessionURL: URL? = nil,
        effort: String? = nil,
        startedAt: Date? = nil
    ) {
        self.id = id
        self.provider = provider
        self.title = title
        self.detail = detail
        self.progress = progress
        self.status = status
        self.updatedAt = updatedAt
        self.sessionURL = sessionURL
        self.effort = effort
        self.startedAt = startedAt
    }
}

/// A daily AI usage aggregate. `cost` and `model` are optional.
public struct AIUsageSample: Codable, Equatable, Sendable {
    /// Stable identifier for an internally maintained daily aggregate.
    public var sourceID: String?
    public var provider: AIProvider
    public var timestamp: Date
    public var inputTokens: Int
    public var outputTokens: Int
    public var costUSD: Double?
    public var model: String?

    public init(
        sourceID: String? = nil,
        provider: AIProvider,
        timestamp: Date,
        inputTokens: Int,
        outputTokens: Int,
        costUSD: Double? = nil,
        model: String? = nil
    ) {
        self.sourceID = sourceID
        self.provider = provider
        self.timestamp = timestamp
        self.inputTokens = inputTokens
        self.outputTokens = outputTokens
        self.costUSD = costUSD
        self.model = model
    }

    public var totalTokens: Int { inputTokens + outputTokens }
}

public enum NoticeKind: String, Codable, Sendable {
    case info
    case success
    case warning
    case error

    public init?(token: String) {
        self.init(rawValue: token.lowercased())
    }
}

public enum NoticeSide: String, Codable, Sendable {
    case left
    case right

    public init?(token: String) {
        self.init(rawValue: token.lowercased())
    }
}

/// Presentation style for regular, message, and system-status notices; old state files decode absent values as `.standard`.
public enum NoticeStyle: String, Codable, Sendable {
    case standard
    case message
    case status
    case headphone
}

/// A single battery reading in an accessory notice. `level == nil` means the system did not provide a charge level for that device.
public struct NoticeBatteryLevel: Codable, Equatable, Sendable, Identifiable {
    public var label: String
    public var level: Int?

    public init(label: String, level: Int?) {
        self.label = label
        self.level = level.map { min(max($0, 0), 100) }
    }

    public var id: String { label }
}

/// Transient notice strip shown on either side of the island.
public struct IslandNotice: Codable, Equatable, Sendable, Identifiable {
    public var id: String
    public var title: String
    public var detail: String?
    public var kind: NoticeKind
    public var side: NoticeSide
    public var createdAt: Date
    public var progress: Double?
    /// Optional media artwork used by the compact playback wing. It is not
    /// required for AI or ordinary notices and remains absent in old state files.
    public var artworkData: Data?
    /// Presentation style; defaults to `.standard` when absent.
    public var style: NoticeStyle
    /// Source app display name (meaningful only when `style == .message`).
    public var appName: String?
    /// Optional bundle ID used to resolve the installed app's icon.
    public var appBundleIdentifier: String?
    /// SF Symbol name used on the left side of a status notice.
    public var symbolName: String?
    /// Individual battery readings for attached accessories; when absent, preserves the existing display and decoding behaviour for old notices.
    public var batteryLevels: [NoticeBatteryLevel]?

    public init(
        id: String = UUID().uuidString,
        title: String,
        detail: String? = nil,
        kind: NoticeKind = .info,
        side: NoticeSide = .right,
        createdAt: Date = Date(),
        progress: Double? = nil,
        artworkData: Data? = nil,
        style: NoticeStyle = .standard,
        appName: String? = nil,
        appBundleIdentifier: String? = nil,
        symbolName: String? = nil,
        batteryLevels: [NoticeBatteryLevel]? = nil
    ) {
        self.id = id
        self.title = title
        self.detail = detail
        self.kind = kind
        self.side = side
        self.createdAt = createdAt
        self.progress = progress.map { min(max($0, 0), 1) }
        self.artworkData = artworkData
        self.style = style
        self.appName = appName
        self.appBundleIdentifier = appBundleIdentifier
        self.symbolName = symbolName
        self.batteryLevels = batteryLevels
    }

    private enum CodingKeys: String, CodingKey {
        case id, title, detail, kind, side, createdAt, progress, artworkData
        case style, appName, appBundleIdentifier, symbolName, batteryLevels
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        title = try container.decode(String.self, forKey: .title)
        detail = try container.decodeIfPresent(String.self, forKey: .detail)
        kind = try container.decode(NoticeKind.self, forKey: .kind)
        side = try container.decode(NoticeSide.self, forKey: .side)
        createdAt = try container.decode(Date.self, forKey: .createdAt)
        progress = try container.decodeIfPresent(Double.self, forKey: .progress)
        artworkData = try container.decodeIfPresent(Data.self, forKey: .artworkData)
        style = try container.decodeIfPresent(NoticeStyle.self, forKey: .style) ?? .standard
        appName = try container.decodeIfPresent(String.self, forKey: .appName)
        appBundleIdentifier = try container.decodeIfPresent(String.self, forKey: .appBundleIdentifier)
        symbolName = try container.decodeIfPresent(String.self, forKey: .symbolName)
        batteryLevels = try container.decodeIfPresent([NoticeBatteryLevel].self, forKey: .batteryLevels)
    }
}

/// Testable model for an inbound IPC message notification: produces a pair of left/right `IslandNotice` values in one shot.
public struct MessageNotification: Equatable, Sendable {
    public static let maxContentLength = 48

    public var appName: String
    public var sender: String
    public var content: String
    public var appBundleIdentifier: String?
    public var createdAt: Date
    public var pairID: String

    public init(
        appName: String,
        sender: String,
        content: String,
        appBundleIdentifier: String? = nil,
        createdAt: Date = Date(),
        pairID: String = UUID().uuidString
    ) {
        self.appName = appName
        self.sender = sender
        self.content = Self.normalizeContent(content)
        self.appBundleIdentifier = appBundleIdentifier.flatMap { value in
            let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
            return trimmed.isEmpty ? nil : trimmed
        }
        self.createdAt = createdAt
        self.pairID = pairID
    }

    /// Collapses whitespace/newlines and truncates to the short display length, appending an ellipsis when over the limit.
    public static func normalizeContent(_ raw: String, maxLength: Int = maxContentLength) -> String {
        let parts = raw.split { $0.isWhitespace || $0.isNewline }.filter { !$0.isEmpty }
        let collapsed = parts.joined(separator: " ")
        guard collapsed.count > maxLength else { return collapsed }
        let end = collapsed.index(collapsed.startIndex, offsetBy: maxLength)
        return String(collapsed[..<end]) + "…"
    }

    /// Left side: app + sender; right side: message body. Both share the same `pairID` prefix and timestamp.
    public func makeNotices() -> (left: IslandNotice, right: IslandNotice) {
        let left = IslandNotice(
            id: "message-\(pairID)-left",
            title: sender,
            detail: appName,
            kind: .info,
            side: .left,
            createdAt: createdAt,
            style: .message,
            appName: appName,
            appBundleIdentifier: appBundleIdentifier
        )
        let right = IslandNotice(
            id: "message-\(pairID)-right",
            title: content,
            detail: nil,
            kind: .info,
            side: .right,
            createdAt: createdAt,
            style: .message,
            appName: appName,
            appBundleIdentifier: appBundleIdentifier
        )
        return (left, right)
    }
}

/// Persisted aggregate state.
public struct AIState: Codable, Equatable, Sendable {
    public var tasks: [AIProgressTask]
    public var usageSamples: [AIUsageSample]
    public var notices: [IslandNotice]

    public init(
        tasks: [AIProgressTask] = [],
        usageSamples: [AIUsageSample] = [],
        notices: [IslandNotice] = []
    ) {
        self.tasks = tasks
        self.usageSamples = usageSamples
        self.notices = notices
    }

    public static let empty = AIState()
}

File: windows/Core/src/AIModels.cpp
#include "zisla/core/AIModels.hpp"

#include <array>
#include <limits>
#include <utility>

namespace zisla::core {
namespace {

struct ProviderAlias {
    std::string_view token;
    AIProvider provider;
};

constexpr auto provider_aliases = std::to_array<ProviderAlias>({
    {"claude", AIProvider::claude},
    {"claude-code", AIProvider::claude},
    {"claude-cli", AIProvider::claude},
    {"claude-desktop", AIProvider::claude},
    {"anthropic-claude", AIProvider::claude},
    {"codex", AIProvider::codex},
    {"openai-codex", AIProvider::codex},
    {"codex-cli", AIProvider::codex},
    {"codex-desktop", AIProvider::codex},
    {"codex-app", AIProvider::codex},
    {"gemini", AIProvider::gemini},
    {"google-gemini", AIProvider::gemini},
    {"gemini-cli", AIProvider::gemini},
    {"gemini-code-assist", AIProvider::gemini},
    {"grok", AIProvider::grok},
    {"grok-cli", AIProvider::grok},
    {"xai", AIProvider::grok},
    {"x-ai", AIProvider::grok},
    {"gpt", AIProvider::gpt},
    {"openai", AIProvider::gpt},
    {"chatgpt", AIProvider::gpt},
    {"chat-gpt", AIProvider::gpt},
    {"openai-gpt", AIProvider::gpt},
    {"copilot", AIProvider::copilot},
    {"github-copilot", AIProvider::copilot},
    {"github copilot", AIProvider::copilot},
    {"copilot-cli", AIProvider::copilot},
    {"copilot-chat", AIProvider::copilot},
    {"github.copilot-chat", AIProvider::copilot},
    {"kimi", AIProvider::kimi},
    {"kimi-code", AIProvider::kimi},
    {"kimi-code-cli", AIProvider::kimi},
    {"kimi-vscode", AIProvider::kimi},
    {"moonshot-kimi", AIProvider::kimi},
    {"moonshot-ai.kimi-code", AIProvider::kimi},
    {"qwen", AIProvider::qwen},
    {"tongyi", AIProvider::qwen},
    {"qwen-code", AIProvider::qwen},
    {"qwen-code-cli", AIProvider::qwen},
    {"qwen-vscode", AIProvider::qwen},
    {"coder", AIProvider::coder},
    {"qwen-coder", AIProvider::coder},
    {"qoder", AIProvider::coder},
    {"qoder-cli", AIProvider::coder},
    {"qoderwork", AIProvider::coder},
    {"qoder-work", AIProvider::coder},
    {"qoderworkcn", AIProvider::coder},
    {"qoderwork-cn", AIProvider::coder},
    {"qoderwork cn", AIProvider::coder},
    {"qoderwake", AIProvider::coder},
    {"qoder-wake", AIProvider::coder},
    {"trae", AIProvider::trae},
    {"trae-work", AIProvider::trae},
    {"traework", AIProvider::trae},
    {"trae-work-cn", AIProvider::trae},
    {"trae-solo", AIProvider::trae},
    {"trae-solo-cn", AIProvider::trae},
    {"trae-cn", AIProvider::trae},
    {"opencode", AIProvider::opencode},
    {"open-code", AIProvider::opencode},
    {"open_code", AIProvider::opencode},
    {"harness", AIProvider::harness},
    {"harnext", AIProvider::harness},
    {"harnext-cli", AIProvider::harness},
    {"harness-cli", AIProvider::harness},
    {"doubao", AIProvider::doubao},
    {"\xE8\xB1\x86\xE5\x8C\x85", AIProvider::doubao},
});

constexpr char ascii_lower(char value) noexcept {
    return value >= 'A' && value <= 'Z'
        ? static_cast<char>(value + ('a' - 'A'))
        : value;
}

bool ascii_case_equal(std::string_view lhs, std::string_view rhs) noexcept {
    if (lhs.size() != rhs.size()) {
        return false;
    }
    for (std::size_t index = 0; index < lhs.size(); ++index) {
        if (ascii_lower(lhs[index]) != ascii_lower(rhs[index])) {
            return false;
        }
    }
    return true;
}

constexpr bool ascii_whitespace(unsigned char value) noexcept {
    return value == ' '
        || value == '\t'
        || value == '\n'
        || value == '\r'
        || value == '\f'
        || value == '\v';
}

std::size_t utf8_prefix_bytes(
    std::string_view value,
    std::size_t maximum_code_points) noexcept {
    std::size_t code_points = 0;
    for (std::size_t index = 0; index < value.size(); ++index) {
        const auto byte = static_cast<unsigned char>(value[index]);
        if ((byte & 0xC0U) == 0x80U) {
            continue;
        }
        if (code_points == maximum_code_points) {
            return index;
        }
        ++code_points;
    }
    return value.size();
}

}  // namespace

std::optional<AIProvider> parse_ai_provider(std::string_view token) noexcept {
    for (const auto& alias : provider_aliases) {
        if (ascii_case_equal(token, alias.token)) {
            return alias.provider;
        }
    }
    return std::nullopt;
}

std::string_view ai_provider_token(AIProvider provider) noexcept {
    switch (provider) {
    case AIProvider::claude: return "claude";
    case AIProvider::codex: return "codex";
    case AIProvider::gemini: return "gemini";
    case AIProvider::grok: return "grok";
    case AIProvider::gpt: return "gpt";
    case AIProvider::copilot: return "copilot";
    case AIProvider::kimi: return "kimi";
    case AIProvider::qwen: return "qwen";
    case AIProvider::coder: return "coder";
    case AIProvider::trae: return "trae";
    case AIProvider::opencode: return "opencode";
    case AIProvider::harness: return "harness";
    case AIProvider::doubao: return "doubao";
    }
    return {};
}

std::optional<NoticeKind> parse_notice_kind(std::string_view token) noexcept {
    if (ascii_case_equal(token, "info")) return NoticeKind::info;
    if (ascii_case_equal(token, "success")) return NoticeKind::success;
    if (ascii_case_equal(token, "warning")) return NoticeKind::warning;
    if (ascii_case_equal(token, "error")) return NoticeKind::error;
    return std::nullopt;
}

std::string_view notice_kind_token(NoticeKind kind) noexcept {
    switch (kind) {
    case NoticeKind::info: return "info";
    case NoticeKind::success: return "success";
    case NoticeKind::warning: return "warning";
    case NoticeKind::error: return "error";
    }
    return {};
}

std::optional<NoticeSide> parse_notice_side(std::string_view token) noexcept {
    if (ascii_case_equal(token, "left")) return NoticeSide::left;
    if (ascii_case_equal(token, "right")) return NoticeSide::right;
    return std::nullopt;
}

std::string_view notice_side_token(NoticeSide side) noexcept {
    switch (side) {
    case NoticeSide::left: return "left";
    case NoticeSide::right: return "right";
    }
    return {};
}

std::optional<NoticeStyle> parse_notice_style(std::string_view token) noexcept {
    if (ascii_case_equal(token, "standard")) return NoticeStyle::standard;
    if (ascii_case_equal(token, "message")) return NoticeStyle::message;
    if (ascii_case_equal(token, "status")) return NoticeStyle::status;
    if (ascii_case_equal(token, "headphone")) return NoticeStyle::headphone;
    return std::nullopt;
}

std::string_view notice_style_token(NoticeStyle style) noexcept {
    switch (style) {
    case NoticeStyle::standard: return "standard";
    case NoticeStyle::message: return "message";
    case NoticeStyle::status: return "status";
    case NoticeStyle::headphone: return "headphone";
    }
    return {};
}

std::string MessageNotification::normalize_content(
    std::string_view raw,
    std::size_t maximum_length) {
    std::string collapsed;
    collapsed.reserve(raw.size());
    bool pending_space = false;
    for (const auto value : raw) {
        if (ascii_whitespace(static_cast<unsigned char>(value))) {
            pending_space = !collapsed.empty();
            continue;
        }
        if (pending_space) {
            collapsed.push_back(' ');
            pending_space = false;
        }
        collapsed.push_back(value);
    }

    const auto prefix_size = utf8_prefix_bytes(collapsed, maximum_length);
    if (prefix_size == collapsed.size()) {
        return collapsed;
    }
    collapsed.resize(prefix_size);
    collapsed += "\xE2\x80\xA6";
    return collapsed;
}

std::pair<IslandNotice, IslandNotice> MessageNotification::make_notices() const {
    IslandNotice left{
        .id = "message-" + pair_id + "-left",
        .title = sender,
        .detail = app_name,
        .kind = NoticeKind::info,
        .side = NoticeSide::left,
        .created_at_unix_ms = created_at_unix_ms,
        .style = NoticeStyle::message,
        .app_name = app_name,
        .app_bundle_identifier = app_bundle_identifier,
    };
    IslandNotice right{
        .id = "message-" + pair_id + "-right",
        .title = normalize_content(content),
        .kind = NoticeKind::info,
        .side = NoticeSide::right,
        .created_at_unix_ms = created_at_unix_ms,
        .style = NoticeStyle::message,
        .app_name = app_name,
        .app_bundle_identifier = app_bundle_identifier,
    };
    return {std::move(left), std::move(right)};
}

std::optional<AIProgressStatus> parse_ai_progress_status(
    std::string_view token) noexcept {
    if (ascii_case_equal(token, "queued")) return AIProgressStatus::queued;
    if (ascii_case_equal(token, "running")) return AIProgressStatus::running;
    if (ascii_case_equal(token, "blocked")) return AIProgressStatus::blocked;
    if (ascii_case_equal(token, "error")) return AIProgressStatus::error;
    if (ascii_case_equal(token, "succeeded")) return AIProgressStatus::succeeded;
    if (ascii_case_equal(token, "failed")) return AIProgressStatus::failed;
    return std::nullopt;
}

std::string_view ai_progress_status_token(AIProgressStatus status) noexcept {
    switch (status) {
    case AIProgressStatus::queued: return "queued";
    case AIProgressStatus::running: return "running";
    case AIProgressStatus::blocked: return "blocked";
    case AIProgressStatus::error: return "error";
    case AIProgressStatus::succeeded: return "succeeded";
    case AIProgressStatus::failed: return "failed";
    }
    return {};
}

bool is_active(AIProgressStatus status) noexcept {
    switch (status) {
    case AIProgressStatus::queued:
    case AIProgressStatus::running:
    case AIProgressStatus::blocked:
    case AIProgressStatus::error:
        return true;
    case AIProgressStatus::succeeded:
    case AIProgressStatus::failed:
        return false;
    }
    return false;
}

NoticeKind notice_kind_for(AIProgressStatus status) noexcept {
    switch (status) {
    case AIProgressStatus::queued:
    case AIProgressStatus::running:
        return NoticeKind::info;
    case AIProgressStatus::blocked:
        return NoticeKind::warning;
    case AIProgressStatus::error:
    case AIProgressStatus::failed:
        return NoticeKind::error;
    case AIProgressStatus::succeeded:
        return NoticeKind::success;
    }
    return NoticeKind::error;
}

std::uint64_t AIUsageSample::total_tokens() const noexcept {
    const auto remaining = std::numeric_limits<std::uint64_t>::max() - input_tokens;
    return output_tokens > remaining
        ? std::numeric_limits<std::uint64_t>::max()
        : input_tokens + output_tokens;
}

}  // namespace zisla::core
